### PR TITLE
[libcxx] Improve libcxx tests when using Optimizations.

### PR DESCRIPTION
**Pull request by Stylie777**
Within the libcxx tests, some tests compare a pointer to a Global Variable to one that is stored locally. At high levels of optmization the pointers are the same, but the assert is failing as the compiler cannot properly compare them. 

This changes how these variables are defined in specific tests to ensure the asserts pass. This is done by 
1. Updating a variant of `DoNotOptimize` used to accept L value references. This is a different variant that is currently used to ensure that the compiler cannot follow the data-flow through it by using both the input and output from the inline asm
2. Add in missing `DoNotOptimize` calls to tests. When using -O2 optimisations, these are failing because the pointer comparison fails when both values are the same. Adding the function call fixes this.

No change to the tests functionality is included.

### CODE REVIEW

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp [hunk 1]
```diff
@@ -51,7 +51,8 @@ int main(int, char**) {
     // Test with an overaligned type
     {
         new_called = delete_called = 0;
-        OverAligned* x = DoNotOptimize(new OverAligned[3]);
+        OverAligned* dummy_data_block = new (std::nothrow) OverAligned;
```
Comment (ldionne) [inline libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:54]: This change uses the nothrow variant of `operator new`, but that is not what we are trying to test here.

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp [hunk 2]
```diff
@@ -27,16 +27,31 @@
 #include "test_macros.h"
 #include "../types.h"
 
+class Data{
```
Comment (ldionne) [inline libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp:30]: I don't understand why this change is necessary.

**Review by ldionne — CHANGES_REQUESTED** [review]
Sorry if I am missing something obvious, but I basically don't understand this change. You mention that "the compiler cannot properly compare the values", what do you mean by that?

What issue are you encountering and with what configuration? Can you share a reproducer showcasing the issue?

**Comment by Stylie777** [comment]
> Sorry if I am missing something obvious, but I basically don't understand this change. You mention that "the compiler cannot properly compare the values", what do you mean by that?

@ldionne  Apologies, not sure I explained it very well there. 

The issue that has been observed is that with certain configurations when using C++17, when comparing the pointers of `x` and `DummyData`, they result was coming out as false when the pointers are identical (see below). This was occurring on 6 tests, 3 of which can be remedied by utilising a different `DoNotOptimize` function that uses the input and output from the inline asm. The other three I used the class to define the DummyData, as then this removes the issues that can arise from comparing a global pointer to that of a local one. 

Example:
```
0000000000060940 <- `x`
0000000000060940 <- `DummyData`
0 <- (x == DummyData)
```

This is running on AArch64 V8 with the following compile options: `--target=aarch64-none-elf -march=armv8 -std=c++17 -O2`

**Review by jthackray — APPROVED** [review]
LGTM, once you've clang-formatted the code.

**Comment by Stylie777** [comment]
Updated. Ran Clang-Format

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp [hunk 3]
```diff
@@ -51,7 +51,8 @@ int main(int, char**) {
     // Test with an overaligned type
     {
         new_called = delete_called = 0;
-        OverAligned* x = DoNotOptimize(new OverAligned[3]);
+        OverAligned* dummy_data_block = new OverAligned;
```
Comment (ldionne) [inline libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:54]: This is still incorrect. This test is trying to test `operator new[]`, but you are not calling it anymore. You are calling the non-array version.

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp [hunk 4]
```diff
@@ -27,16 +27,28 @@
...
+public:
+  Data()  = default;
+  ~Data() = default;
+
+  char* getDummyData() { return this->dummy_data_; }
+
+  std::size_t getDummyDataSize() { return sizeof(this->dummy_data_); }
+
+private:
+  alignas(OverAligned) char dummy_data_[alignof(OverAligned) * 3];
```
Comment (ldionne) [inline libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp:40]: I don't understand why indirecting through a class solves anything here. After all you're still returning the address of a global variable.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp (hunk 4 above)
Comment (Stylie777) [reply libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.pass.cpp]: I have removed the classes and reverted to the initial behaviour. Please see the initial comment as the changes made are different to those previously reviewed. Thanks

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_align_nothrow.replace.pass.cpp [hunk 5]
```diff
@@ -47,7 +47,7 @@ int main(int, char**) {
     // Test with an overaligned type
     {
         new_nothrow_called = delete_called = 0;
-        OverAligned* x = new (std::nothrow) OverAligned;
+        OverAligned* x                     = DoNotOptimize(new (std::nothrow) OverAligned);
```
Comment (ldionne) [inline libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_align_nothrow.replace.pass.cpp:50]: Ah! I understand this part of the change -- that looks like an oversight to me. However, I don't understand why you need to add a rvalue-reference overload to the function below, it seems to me that the `const&` overload should handle this correctly, no?

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_align_nothrow.replace.pass.cpp (hunk 5 above)
Comment (Stylie777) [reply libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.single/new.size_align_nothrow.replace.pass.cpp]: The change to `&&value` was not the correct change, so have reverted this back to `&value` for `DoNotOptimize`.

In these test cases, we have found that the `const&` overload was not allowing for comparison of the pointers correctly, even when the values were the same. See explanation from @ostannard below:

"think that's because DoNotOptimize only uses the value as an input. This prevents the compiler from removing and data-flow leading up to it, but it can still "see" that one pointer came from new, and the other came from the address of a global, so they "can't" alias
This works for me, using both input and output from the inline asm, so the compiler can't follow the data-flow through it:"
```
template <class Tp>
inline Tp DoNotOptimize(Tp value) {
    asm volatile("" : "+r,m"(value) : : "memory");
    return value;
}
```

I have now scaled back the changes to only apply to the tests rather than change the function itself.

File: libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp [hunk 6]
```diff
@@ -51,7 +51,8 @@ int main(int, char**) {
     // Test with an overaligned type
     {
         new_called = delete_called = 0;
-        OverAligned* x = DoNotOptimize(new OverAligned[3]);
+        OverAligned* dummy_data_block = new OverAligned[3];
```
[the remark this answers is not in the document]
Comment (Stylie777) [reply libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:54]: Yes we do - hence the change in these tests to a different variant of the `DoNotOptimize` function.

[the remark this answers is not in the document]
Comment (Stylie777) [reply libcxx/test/std/language.support/support.dynamic/new.delete/new.delete.array/new.size_align.replace.indirect.pass.cpp:54]: Updated

**Review by ldionne — APPROVED** [review]

**Comment by Stylie777** [comment]
Thanks for merging @ldionne and thank you for the review. Much appreciated!
